lib/rust/ensogl/core/src/display/object/event.rs

```diff
@@ -0,0 +1,175 @@
+//! Events implementation. Events behave in a similar way to JavaScript Events. When an event is
+//! emitted, it is propagated in three stages: capturing, target, and bubbling. Each stage is
+//! configurable and some events propagation can be cancelled. To learn more about the mechanics,
+//! see: https://javascript.info/bubbling-and-capturing.
+
+use crate::prelude::*;
+
+use super::class::Instance;
+use super::class::WeakInstance;
+
+
+
+// =============
+// === State ===
+// =============
+
+/// Event state. It can be used to determine whether the event is being propagated, its propagation
+/// is cancelled, or that the propagation cannot be cancelled. See docs of this module to learn
+/// more.
+#[allow(missing_docs)]
+#[derive(Debug, Clone, Copy, Default, PartialEq, Eq)]
+pub enum State {
+    #[default]
+    Running,
+    RunningNonCancellable,
+    Cancelled,
+}
+
+
+
+// =================
+// === SomeEvent ===
+// =================
+
+/// A counterpart of [`Event`] with hidden payload type. It is used to construct, configure, and
+/// emit new events.
+#[allow(missing_docs)]
+#[derive(Clone, CloneRef, Debug)]
+pub struct SomeEvent {
+    pub data:     frp::AnyData,
+    state:        Rc<Cell<State>>,
+    pub captures: Rc<Cell<bool>>,
+    pub bubbles:  Rc<Cell<bool>>,
```

Comment:
What are those fields? Do they mean "I'm in capture/bubble stage" or maybe "I go the capture/bubble state"?

And why are they only here, not in Event (which is called a "counterpart").

Reply:
1. They mean the same as in JS - this is the current stage. I'll add docs.
2. Yeah, the word "counterpart" is bad there, I'll refactor it.